crates/ext-processor/src/service.rs

```diff
@@ -0,0 +1,1045 @@
+                shared_params.clone(),
+                http_req.clone(),
+            )?;
+
+            plugin_instances.push(Arc::new(Mutex::new(PluginInstance::new(
+                plugin.clone(),
+                request_context,
+            )?)));
+        }
+        Ok(plugin_instances)
+    }
+
+    async fn execute_request_phase(
+        plugin_instances: Vec<Arc<Mutex<PluginInstance>>>,
+        timeout_duration: std::time::Duration,
+    ) -> DecisionComponents {
+        Self::execute_request_phase_one(plugin_instances.clone(), timeout_duration).await;
+        Self::execute_request_phase_two(plugin_instances.clone(), timeout_duration).await
+    }
+
+    async fn execute_request_phase_one(
+        plugin_instances: Vec<Arc<Mutex<PluginInstance>>>,
+        timeout_duration: std::time::Duration,
+    ) {
+        let mut phase_one_tasks = JoinSet::new();
+        for plugin_instance in plugin_instances.clone() {
+            let phase_one_child_span = tracing::info_span!("execute on_request",);
+            phase_one_tasks.spawn(
+                timeout(timeout_duration, async move {
+                    // TODO: avoid unwraps
+                    Self::execute_plugin_initialization(plugin_instance.clone()).unwrap();
+                    Self::execute_on_request(plugin_instance.clone()).unwrap();
+                })
+                .instrument(phase_one_child_span.or_current()),
+            );
+        }
+        // efficiently hand execution off to the plugins
+        tokio::task::yield_now().await;
+
+        while let Some(r) = phase_one_tasks.join_next().await {
+            match r {
+                Ok(Ok(_)) => {}
+                Ok(Err(e)) => {
+                    warn!(
+                        message = "timeout on plugin execution",
+                        elapsed = ?e,
+                    );
+                }
+                Err(e) => {
+                    warn!(
+                        message = "join error on plugin execution",
+                        error_message = ?e,
+                    );
+                }
+            }
+        }
+    }
+
+    async fn execute_request_phase_two(
+        plugin_instances: Vec<Arc<Mutex<PluginInstance>>>,
+        timeout_duration: std::time::Duration,
+    ) -> DecisionComponents {
+        let decision_components = Arc::new(Mutex::new(Vec::with_capacity(plugin_instances.len())));
+        let mut phase_two_tasks = JoinSet::new();
+        for plugin_instance in plugin_instances.clone() {
+            let phase_two_child_span = tracing::info_span!("execute on_request_decision",);
+            let decision_components = decision_components.clone();
+            phase_two_tasks.spawn(
+                timeout(timeout_duration, async move {
+                    let decision_result =
+                        Self::execute_on_request_decision(plugin_instance.clone());
+                    let mut decision_component = decision_result.unwrap();
```

Comment:
Possibly. I need to take a look over the decision ones specifically because there are some cases where it should be guaranteed to never error. Those should have `unwrap`s replaced with `expect` I think.

Reply:
Looks like that's not the case here though. I'm going to get rid of the unwraps.